File: src/test/java/de/pflugradts/passbird/application/util/BytesComparatorTest.java
package de.pflugradts.passbird.application.util;

import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.model.transfer.BytesComparator;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThat;

class BytesComparatorTest {

    private final BytesComparator comparator = new BytesComparator();

    @Nested
    class NullAndEmptyTest {

        @Test
        void shouldCompare_NullAndNull() {
            // given
            final Bytes bytes1 = null;
            final Bytes bytes2 = null;

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isZero();
            assertThat(comparator.compare(bytes2, bytes1)).isZero();
        }

        @Test
        void shouldCompare_NullAndEmpty() {
            // given
            final Bytes bytes1 = Bytes.emptyBytes();
            final Bytes bytes2 = null;

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isZero();
            assertThat(comparator.compare(bytes2, bytes1)).isZero();
        }

        @Test
        void shouldCompare_NullAndNonEmpty() {
            // given
            final Bytes bytes1 = null;
            final Bytes bytes2 = Bytes.bytesOf("1");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_EmptyAndNonEmpty() {
            // given
            final Bytes bytes1 = Bytes.emptyBytes();
            final Bytes bytes2 = Bytes.bytesOf("1");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
       }

    }

    @Nested
    class SymbolsAndDigitsTest {

        @Test
        void shouldCompare_SymbolAndDigit() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("!");
            final Bytes bytes2 = Bytes.bytesOf("1");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_SymbolAndLetter() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("!");
            final Bytes bytes2 = Bytes.bytesOf("a");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_DigitAndLetter() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("1");
            final Bytes bytes2 = Bytes.bytesOf("a");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

    }

    @Nested
    class LettersTest {

        @Test
        void shouldCompare_UppercaseAndLowercase() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("test");
            final Bytes bytes2 = Bytes.bytesOf("TEST");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isZero();
            assertThat(comparator.compare(bytes2, bytes1)).isZero();
        }

        @Test
        void shouldCompare_ShortAndLong() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("test");
            final Bytes bytes2 = Bytes.bytesOf("TESTING");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_LastCharDiffers() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("testingA");
            final Bytes bytes2 = Bytes.bytesOf("testingB");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

    }

    @Nested
    class ArbitraryStringsTest {

        @Test
        void shouldCompare_ArbitraryStrings1() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("BzlPa");
            final Bytes bytes2 = Bytes.bytesOf("NeIKii75");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_ArbitraryStrings2() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("Ue1e;Tv");
            final Bytes bytes2 = Bytes.bytesOf("VknA@");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_ArbitraryStrings3() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("(F");
            final Bytes bytes2 = Bytes.bytesOf("4");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_ArbitraryStrings4() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("789nGE");
            final Bytes bytes2 = Bytes.bytesOf("KjLX%)Dm");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

        @Test
        void shouldCompare_ArbitraryStrings5() {
            // given
            final Bytes bytes1 = Bytes.bytesOf("7g.");
            final Bytes bytes2 = Bytes.bytesOf("r^)HVd%^f");

            // when / then
            assertThat(comparator.compare(bytes1, bytes2)).isNegative();
            assertThat(comparator.compare(bytes2, bytes1)).isPositive();
        }

    }

}

File: src/test/kotlin/de/pflugradts/passbird/domain/model/transfer/BytesComparatorTest.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.transfer.Bytes.Companion.bytesOf
import de.pflugradts.passbird.domain.model.transfer.Bytes.Companion.emptyBytes
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import strikt.api.expectThat
import strikt.assertions.isEqualTo
import strikt.assertions.isGreaterThan
import strikt.assertions.isLessThan

internal class BytesComparatorTest {
    private val comparator = BytesComparator()

    @Test
    fun `should compare empty and non empty`() {
        // given
        val bytes1 = emptyBytes()
        val bytes2 = bytesOf("1")

        // when / then
        expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
        expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
    }

    @Nested
    internal inner class SymbolsAndDigitsTest {
        @Test
        fun `should compare symbol and digit`() {
            // given
            val bytes1 = bytesOf("!")
            val bytes2 = bytesOf("1")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
            expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
        }

        @Test
        fun `should compare symbol and letter`() {
            // given
            val bytes1 = bytesOf("!")
            val bytes2 = bytesOf("a")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
            expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
        }

        @Test
        fun `should compare digit and letter`() {
            // given
            val bytes1 = bytesOf("1")
            val bytes2 = bytesOf("a")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
            expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
        }
    }

    @Nested
    internal inner class LettersTest {
        @Test
        fun `should compare uppercase and lowercase`() {
            // given
            val bytes1 = bytesOf("test")
            val bytes2 = bytesOf("TEST")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isEqualTo 0
            expectThat(comparator.compare(bytes2, bytes1)) isEqualTo 0
        }

        @Test
        fun `should compare short and long`() {
            // given
            val bytes1 = bytesOf("test")
            val bytes2 = bytesOf("TESTING")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
            expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
        }

        @Test
        fun `should compare last char differs`() {
            // given
            val bytes1 = bytesOf("testingA")
            val bytes2 = bytesOf("testingB")

            // when / then
            expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
            expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
        }
    }

    @ParameterizedTest
    @CsvSource(
        "BzlPa,NeIKii75",
        "Ue1e;Tv,VknA@",
        "(F,4",
        "789nGE,KjLX%)Dm",
        "7g.,r^)HVd%^f",
    )
    fun `should compare arbitrary strings`(string1: String, string2: String) {
        // given
        val bytes1 = bytesOf(string1)
        val bytes2 = bytesOf(string2)

        // when / then
        expectThat(comparator.compare(bytes1, bytes2)) isLessThan 0
        expectThat(comparator.compare(bytes2, bytes1)) isGreaterThan 0
    }
}
